Consumer group blocking

bashtanov opened the pull request:
As a preparation to migrate consumer groups, introduce a consumer offsets partition command to block a consumer group.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bashtanov committed:
k/s/group_data_parser: refactor batch type check into a switch

bashtanov committed:
k/s/group_metadata: make kafka::group_metadata_serializer a namespace

kafka::group_metadata_serializer was an base class with a single
implementation, as historically multiple serialization approaches were
used. Currently there's only one, so its member functions do not depend
on the class instance.

Convert it into a namespace, so these functions are not class member
functions anymore. Remove any instantiations of the class.

bashtanov committed:
model/record_batch_types: introduce group_block

for un-/blocking consumer groups in __consumer_offsets topic

bashtanov committed:
k/s/group_data_parser: specify return types for child methods

bashtanov committed:
k/s/group_recovery_consumer,group_tx_tracker_stm: handle group blocking

Handle the group un-/block command. The subject STM and pseudo-STM each
needs to handle them separately to ignore commands coming when blocked
according to its own timeline.

bashtanov committed:
k/s/group_metadata: introduce group_block to store group block command

A type that stores details of this command parsed.

bashtanov committed:
k/s/group_manager: avoid excessive ntp copy

bashtanov committed:
k/s/group_manager: implement blocked groups

A group block does not necessarily refers to an existing group, so store
these separately.

bashtanov committed:
features/table: add consumer_group_migrations feature

As we are adding a new command type for consumer offsets partitions

bashtanov committed:
k/s/tests/consumer_groups: test blocked groups

bashtanov committed:
k/s/group: refactor do_abort_old_txes

Separate the blocking itself into another function so that can be reused
when we need to block them all. Also add a bit of logging.

bashtanov committed:
k/s/tests/consumer_group_recovery: test blocked groups

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--c740d8cd-020e-4828-8667-e1b9dc44af59-->
<!--TEST RESULTS-->
<!--5305c229-7de6-450f-8399-d8bcfe8f079c-->
<!--TEST RESULTS-->
<!--e1fd5967-78de-4835-9409-899af33a1cff-->
<!--TEST RESULTS-->
<!--6a1f3ecf-e352-4645-9c47-62edc477b889-->
<!--TEST RESULTS-->
<!--bf63c311-a11e-4ae2-9af4-dd5a38943872-->
<!--TEST RESULTS-->
<!--70185754-59ce-4f27-972f-fb2895894e5c-->
<!--TEST RESULTS-->
<!--894a9c2f-598f-4f2d-9dad-c44e32e9804b-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67491">build#67491</a></summary>


| test_class                       | test_method                      | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------------|:---------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RaftAvailabilityTest             | test_controller_node_isolation   |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67491#01977f08-c2f2-4f09-b2d1-1fd431862d72 | FLAKY         | 19/21    | upstream reliability is '95.28795811518324'. current run reliability is '90.47619047619048'. drift is 4.81177 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest         | test_node_operations             | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67491#01977f08-c2f2-42af-8749-06dc4e5c3829 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| src/v/storage/opfuzz/opfuzz_test | src/v/storage/opfuzz/opfuzz_test |                                                                                                                                         | unit        | https://buildkite.com/redpanda/redpanda/builds/67491#01977ed3-9759-400b-93a2-06d5b0710d64 | FAIL          | 0/1      |                                                                                                                                                                        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67539">build#67539</a></summary>


| test_class                    | test_method                               | test_arguments                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:------------------------------|:------------------------------------------|:-------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeDelayedEnablementTest | test_enabling_iceberg_in_existing_cluster | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67539#01978280-906f-4931-ab64-3ea66138d86c | FLAKY         | 11/21    | upstream reliability is '84.95575221238938'. current run reliability is '52.38095238095239'. drift is 32.5748 and the allowed drift is set to 50. The test should PASS |
| DatalakeMetricsTest           | test_lag_metrics                          | {"cloud_storage_type": 1}                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67539#01978280-9070-4282-8a30-051b82a3c0cb | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| NodeWiseRecoveryTest          | test_node_wise_recovery                   | {"dead_node_count": 2}                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67539#0197827a-6c5d-4e4b-8242-5a6b01629e9c | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| SimpleEndToEndTest            | test_relaxed_acks                         | {"write_caching": false}                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67539#01978280-9071-4682-91f2-1c7b6abc6dc9 | FLAKY         | 20/21    | upstream reliability is '98.74213836477988'. current run reliability is '95.23809523809523'. drift is 3.50404 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67574">build#67574</a></summary>


| test_class                   | test_method                             | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:-----------------------------|:----------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageTimingStressTest | test_cloud_storage_with_partition_moves | {"cleanup_policy": "compact,delete"}                                                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67574#0197842d-5b1c-46b9-85be-835defb13e1b | FAIL          | 0/1      |                                                                                                                                                           |
| PartitionMoveInterruption    | test_cancelling_partition_move_x_core   | {"compacted": false, "recovery": "restart_recovery", "replication_factor": 1, "unclean_abort": true}                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67574#01978442-a408-4e51-a3e3-899624c5983a | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest     | test_node_operations                    | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67574#0197842d-5b1c-46b9-85be-835defb13e1b | FAIL          | 0/1      |                                                                                                                                                           |
| RandomNodeOperationsTest     | test_node_operations                    | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67574#0197842d-5b1c-46b9-85be-835defb13e1b | FAIL          | 0/1      |                                                                                                                                                           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67652">build#67652</a></summary>


| test_class               | test_method              | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:-------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CompactionGapsTest       | test_translation_no_gaps | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1}                                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67652#01978da3-216a-44e5-af96-d7f6acb06726 | FLAKY         | 19/21    | upstream reliability is '94.71830985915493'. current run reliability is '90.47619047619048'. drift is 4.24212 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations     | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67652#01978da3-216a-44e5-af96-d7f6acb06726 | FLAKY         | 20/21    | upstream reliability is '96.35416666666666'. current run reliability is '95.23809523809523'. drift is 1.11607 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67788">build#67788</a></summary>


| test_class                    | test_method                               | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:------------------------------|:------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeDelayedEnablementTest | test_enabling_iceberg_in_existing_cluster | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67788#0197a1b8-4e53-40ad-9b5b-ad8692a05b1e | FLAKY         | 11/21    | upstream reliability is '72.59684361549498'. current run reliability is '52.38095238095239'. drift is 20.21589 and the allowed drift is set to 50. The test should PASS |
| NodePoolMigrationTest         | test_migrating_redpanda_nodes_to_new_pool | {"balancing_mode": "node_add", "cleanup_policy": "compact,delete", "test_mode": "tiered_storage"}                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67788#0197a1b8-4e51-4b6d-87f7-ebb9d7cc6b84 | FLAKY         | 20/21    |                                                                                                                                                                         |
| RandomNodeOperationsTest      | test_node_operations                      | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": false, "with_iceberg": false}         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67788#0197a1b8-4e51-4b6d-87f7-ebb9d7cc6b84 | FLAKY         | 20/21    |                                                                                                                                                                         |
| RandomNodeOperationsTest      | test_node_operations                      | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67788#0197a1b8-4e4f-40ef-872a-77f3123acec3 | FAIL          | 0/1      |                                                                                                                                                                         |
| TopicDeleteCloudStorageTest   | drop_lifecycle_marker_test                | {"cloud_storage_type": 2}                                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67788#0197a1b5-d766-4222-871b-a8997ff221d7 | FLAKY         | 18/21    | upstream reliability is '96.18834080717488'. current run reliability is '85.71428571428571'. drift is 10.47406 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67921">build#67921</a></summary>


| test_class                       | test_method                | test_arguments           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------------|:---------------------------|:-------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| SimpleEndToEndTest               | test_relaxed_acks          | {"write_caching": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67921#0197a8f6-a760-4e77-9d43-4f1a2ee189eb | FLAKY         | 20/21    | upstream reliability is '99.00990099009901'. current run reliability is '95.23809523809523'. drift is 3.77181 and the allowed drift is set to 50. The test should PASS |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss |                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67921#0197a8f6-a761-49b5-b733-ce891932152a | FLAKY         | 19/21    | upstream reliability is '99.12280701754386'. current run reliability is '90.47619047619048'. drift is 8.64662 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68011">build#68011</a></summary>


| test_class         | test_method       | test_arguments           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------|:------------------|:-------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RpkToolTest        | test_analyze      |                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68011#0197adaf-4f56-472d-9ad1-4e677e89c5f3 | FLAKY         | 20/21    | upstream reliability is '98.03370786516854'. current run reliability is '95.23809523809523'. drift is 2.79561 and the allowed drift is set to 50. The test should PASS |
| SimpleEndToEndTest | test_relaxed_acks | {"write_caching": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68011#0197adb3-aad5-4ad2-b3d9-f78a30dd636c | FLAKY         | 20/21    | upstream reliability is '99.11504424778761'. current run reliability is '95.23809523809523'. drift is 3.87695 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

dotnwat reviewed (commented):
> record_batch_type::group_block

do we need to add this type to the set of offset translator ignored batch types? __consumer_offsets topic is a public kafka topic.

bashtanov commented:
> > record_batch_type::group_block
>
> do we need to add this type to the set of offset translator ignored batch types? __consumer_offsets topic is a public kafka topic.

Yes, that's a good point, I'll do this.

joe-redpanda commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -684,6 +682,84 @@ ss::future<> group_manager::reload_groups() {
+
+    if (blocked) {
+        auto last_error = cluster::tx::errc::none;
+        co_await ss::max_concurrent_for_each(
```
if there's multiple errors last error is only going to get one of them. Do we care about multiple errors for logging?

bashtanov replied:
They will all be logged in `group::try_abort_old_tx` apart from a situation where we cannot grab catchup_lock in shared fashion due to ongoing recovery. I added some logging for this case.

joe-redpanda commented on src/v/kafka/server/group_metadata.h:
```diff
@@ -177,53 +177,29 @@ inline group_metadata_version read_metadata_version(protocol::decoder& reader) {
+    explicit group_block(model::record record);
+    void add_to_batch_builder(storage::record_batch_builder&) const;
 };
+namespace group_metadata_serializer {
```
Any reason to make this a namespace and not a class/struct with statics?

The template system isn't the most kind on namespaces.
https://godbolt.org/z/TGjbqxf49

bashtanov replied:
I think it's more likely that we'll need to avoid repeating the namespace name with `using namespace group_metadata_serializer;` than we'd like to use it as a template parameter.

joe-redpanda commented on src/v/kafka/server/tests/consumer_groups_test.cc:
```diff
@@ -140,6 +141,58 @@ FIXTURE_TEST(empty_offset_commit_request, consumer_offsets_fixture) {
+    kafka::group_instance_id gi("group-instance");
+
+    scoped_config cfg;
+    // cfg.get("group_topic_partitions").set_value(1);
```
intentional?

bashtanov replied:
oops, no, will fix

joe-redpanda reviewed (dismissed):
added nits, otherwise lgtm

bharathv commented on src/v/kafka/server/group_tx_tracker_stm.h:
```diff
@@ -154,6 +156,7 @@ class group_tx_tracker_stm final
     void maybe_end_tx(kafka::group_id, model::producer_identity, model::offset);
 
     all_txs_t _all_txs;
+    chunked_hash_set<kafka::group_id> _blocked_groups;
```
this should go into the snapshot too?

bharathv replied:
missed this? I think this is needed for correctness. There is a local snapshot optimization for this stm that must capture the blocked_groups until snapshot offset.

bharathv commented on src/v/kafka/server/group_data_parser.h:
```diff
@@ -175,43 +199,68 @@ class group_data_parser {
     ss::future<> handle_tx_offsets(
       model::record_batch_header header,
       kafka::group_tx::offsets_metadata data) {
+        if (is_group_blocked_verbose(data.group_id, "tx offsets")) {
```
I'm wondering if we can bump the log level here.. a new transaction after a group block is a bug for sure?

bharathv commented on src/v/kafka/server/group.h:
```diff
@@ -720,6 +720,8 @@ class group final : public ss::enable_lw_shared_from_this<group> {
 
     void set_lag_metrics(consumer_lag_metrics lag_metrics);
 
+    ss::future<cluster::tx::errc> abort_txes(bool expired_only);
```
nit: document "expired only", especially what happens if its false.

bharathv commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -682,6 +682,84 @@ ss::future<> group_manager::reload_groups() {
+              if (it == _groups.end()) {
+                  return ss::make_ready_future<>();
+              }
+              return it->second->abort_txes(true).then(
```
hmmm.. what if block replication succeeds and abort fails? Wondering if we should reverse these operations.. (perhaps cache term before doing it to avoid ABA leadership problems)

edit: looks like attached_partition.term is already cached, so that takes care of fencing on term changes.

wait until all open transactions are aborted, ensure no open transaction (has_transactions_in_progress()), then replicate block.

bashtanov replied:
I wouldn't like to introduce a transaction whack-a-mole loop as it may never end if clients start them eagerly. Instead, my plan is to block groups early but revert it if either transactions fails to abort or command fails to replicate.

bharathv commented on src/v/kafka/server/group_tx_tracker_stm.cc:
```diff
@@ -268,6 +271,18 @@ ss::future<> group_tx_tracker_stm::handle_version_fence(
 }
 
+void group_tx_tracker_stm::handle_group_block(kafka::group_block gb) {
+    if (gb.is_blocked) {
```
I'm wondering if this should also erase any transaction state just to be on the safe side, like how tombstone handles it

IIUC this shouldn't happen in theory because a block is only applied after aborting any inflight transactions but if there is a bug, this might end up blocking compaction.

```
void group_tx_tracker_stm::handle_group_metadata(group_metadata_kv md) {
    if (md.value) {
        vlog(cg_klog.trace, "[group: {}] update", md.key.group_id);
        // A group may checkpoint periodically as the member's state changes,
        // here we retain the group state if the group already exists.
        _all_txs.try_emplace(md.key.group_id, per_group_state{});
    } else {
        vlog(cg_klog.trace, "[group: {}] tombstone", md.key.group_id);
        // A tombstone indicates all the group state can be purged and
        // any transactions can be ignored. Although care must be taken
        // to ensure there are no open transactions before tombstoning
        // a group in the main state machine.
        _all_txs.erase(md.key.group_id);
    }
}
```

bashtanov replied:
Yeah sounds reasonable, I added this. If we trusted it to abort transactions and prevent new ones reliably we could make this STM ignore these batches though.

bharathv replied:
correct.

bharathv commented on src/v/kafka/server/tests/consumer_groups_test.cc:
```diff
@@ -140,6 +141,56 @@ FIXTURE_TEST(empty_offset_commit_request, consumer_offsets_fixture) {
+
+    BOOST_REQUIRE(commit_offset());
+    set_blocked(true);
+    BOOST_REQUIRE(!commit_offset());
```
leadership change to cover group_recovery_consumer code paths? (the new leader should also see the blocked group)

bashtanov replied:
this fixture only runs one broker, but a node restart covers this as well

bharathv reviewed (commented):
guess ducktape tests come in a separate PR that integrates blocking with rest of the unmount machinery?

bashtanov committed:
container/fragmented_vector: add a constructor that copies from a range

Using std::from_range_t tag argument to discourage accidental misuse and
to allow std::ranges::to<> use it

bharathv commented on src/v/container/fragmented_vector.h:
```diff
@@ -120,17 +120,29 @@ class fragmented_vector {
     template<typename Range>
     requires std::ranges::sized_range<Range>
-    explicit fragmented_vector(Range range)
+    explicit fragmented_vector(Range&& range)
```
umm, maybe I'm missing something, this is a forwarding reference, right? 

If the constructor is called with a non const lvalue Range, it resolves to this overload and it moves out of a lvalue? That is ok?

bashtanov replied:
Thanks. I was assuming the concept restricts `Range` to non-references, but I was wrong. Will fix.

bharathv replied:
🤔  `is_reference_v<Range>` should evaluate to `true` to for rvalues, this the concept should evaluate to false for existing call sites? I believe some sorcery is happening here, move() is probably generating an xvalue that is satisfying this concept? Is that intentional?

bashtanov replied:
What's happening here is not unlike what's happening when a pair of copy and move constructors without any templates is sufficient for any reasonable calls. A move constructor allows calling with an rvalue expression or an rvalue reference: in the first situation an rvalue reference is bound implicitly, see (3) [here](https://en.cppreference.com/w/cpp/language/reference_initialization.html).

Here the difference is we allow `Range` to be of any range type, not necessarily `fragmented_vector`, but still restrict it to be of a non-reference type so that `Range &&` is not a forwarding reference.

Regarding your specific question I would appreciate if you could clarify a few bits:
1) what kind of call are you concerned about. Does "for rvalues" mean "when called with an rvalue of a range type" or "when called with an rvalue reference to a range"?
2) what "move()" call do you mean. The one 3 lines below or a move call in the caller?

bharathv replied:
I think I understand it, the part I missed is calling this with std::move(T()) results in  Range=T  and not Range=T&& and the concept requirement is satisfied whereas calling with an lvalue ref results in Range = T&.

bashtanov replied:
That's correct. I just found the docs describing the deduction process: [here, adjustments 3 and 4](https://en.cppreference.com/w/cpp/language/template_argument_deduction.html#Deduction_from_a_function_call)

bharathv commented on src/v/kafka/server/group_tx_tracker_stm.h:
```diff
@@ -133,12 +134,14 @@ class group_tx_tracker_stm final
 
 private:
-    static constexpr int8_t supported_local_snapshot_version = 1;
+    static constexpr int8_t supported_local_snapshot_version = 2;
```
this doesn't need a bump I think. (It should only be bumped if the snapshot should be discarded after upgrade)

bashtanov replied:
yeah thanks for the feature flag it cannot happen, I will remove

bharathv commented on src/v/kafka/server/group_tx_tracker_stm.h:
```diff
@@ -133,12 +134,14 @@ class group_tx_tracker_stm final
-      : serde::envelope<snapshot, serde::version<0>, serde::compat_version<0>> {
+      : serde::envelope<snapshot, serde::version<1>, serde::compat_version<0>> {
         all_txs_t transactions;
+        chunked_vector<kafka::group_id> blocked_groups;
```
oh I see this was added here, perhaps squash this commit 79085150a730057104e2e67518878e0ec64c1a68

bashtanov replied:
do you mean combine f8f1341844b3312c03be0ff5e691dccf96229a31 and 79085150a730057104e2e67518878e0ec64c1a68 into a single commit?

bharathv replied:
right, feel free to skip if its extra work, was thinking these two commits belong together since they are adding new state to the stm.

note: there is a TODO in the commit message.

bharathv commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -682,6 +682,97 @@ ss::future<> group_manager::reload_groups() {
+              "Failed to abort transactions for blocked groups: {}",
+              last_error);
+            // revert the blocking operation
+            _blocked_groups.erase(affected_gids.begin(), affected_gids.end());
```
Should this be exception safe? If insert is done and the co_await throws.....

mmaslankaprv replied:
+1

mmaslankaprv commented on src/v/features/feature_table.h:
```diff
@@ -40,6 +40,7 @@ struct feature_table_snapshot;
 enum class feature : std::uint64_t {
     panda_linking_dr = 1ULL << 1U,
     topic_locations_in_outbound_migrations = 1ULL << 2U,
+    consumer_groups_migrations = 1ULL << 7U,
```
why not `<< 3U` ?

bashtanov replied:
To reduce the probability of a non-trivial merge conflict. It's less when everyone chooses a random index. As you can see the list has gaps everywhere, so our past efforts to keep them sequential didn't help when it came to cleaning.

mmaslankaprv commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -682,6 +682,97 @@ ss::future<> group_manager::reload_groups() {
+    if (!p) {
+        vlog(
+          cg_klog.warn,
+          "set blocked request for {} failed - partition not found",
```
nit: maybe add an information that we tried to look for attached partition ?

mmaslankaprv commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -682,6 +682,97 @@ ss::future<> group_manager::reload_groups() {
+        co_return cluster::tx::errc::coordinator_load_in_progress;
+    }
+    p->catchup_lock->read_unlock();
+    auto holder = p->catchup_lock->hold_read_lock();
```
`co_await` ?

bharathv commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -682,6 +682,98 @@ ss::future<> group_manager::reload_groups() {
+        co_return cluster::tx::errc::coordinator_load_in_progress;
+    }
+    p->catchup_lock->read_unlock();
+    auto holder = co_await p->catchup_lock->hold_read_lock();
```
there is a weird existing pattern in the code of group_manager.cc that attempts to `try_take_lock()` in one continuation to fix some historical bug (don't ask me why, haha), so the pattern is something like..

```


if (try_read_lock()) {
  
 .. actualy_take_lock() in the same continuation

}

```

`actualy_take_lock` is usually a continuation so the lock is guaranteed to be acquired  if `try_read_lock()` succeeds but a co_await introduces a scheduling point, probably good to follow that pattern, very unlikely this is going to be a problem since this operation is conditional on some migration but 🤷 

We should probably rewrite `try_read_lock()` to also acquire the lock rather than returning a bool.

bashtanov replied:
There's actually an `rwlock::attempt_read_lock` that does exactly what we need. I'll refactor.

bharathv reviewed (commented):
few nits but lgtm.

bashtanov committed:
k/s/group_manager: refactor to use ssx::rwlock::attempt_read_lock

There was a pattern of checking lock availability, unlocking and then
relocking with units held, all in one scheduling segment.
`ssx::rwlock::attempt_read_lock` allows to do it in one go, greatly
reducing code complexity.

bashtanov commented:
@dotnwat @rockwotj could you have a look at this commit please https://github.com/redpanda-data/redpanda/pull/26489/commits/70468fc9ec6a393b16065d9fe04377493dcc8515? It's adding an explicit constructor to fragmented_vector for it to copy from another range.

StephanDollberg commented on src/v/container/fragmented_vector.h:
```diff
@@ -120,17 +120,29 @@ class fragmented_vector {
-    requires std::ranges::sized_range<Range>
-    explicit fragmented_vector(Range range)
+    requires(std::ranges::sized_range<Range> && !std::is_reference_v<Range>)
+    explicit fragmented_vector(Range&& range)
```
This is effectively a copy constructor?

bashtanov replied:
No, this is not unlike a move constructor, but accepts various types of ranges. The one that is like a copy constructor is below.

StephanDollberg replied:
Yeah sorry I just meant this type of constructor. I think originally we intentionally didn't add a copy constructor (though I guess this one is marked explicit) which is why there is the `copy` method. @rockwotj probably knows the details.

bashtanov replied:
Yeah I'm curious what might be a situation an explicit constructor may still be accidentally misused. If there is one possible I don't mind adding an `std::from_range_t` dummy parameter in front.

rockwotj replied:
It effectively becomes a copy constructor for `chunked_vector` just a lot slower as well.

We've started adding `copy` methods to things to prevent accidental copies. Usually with `chunked_vectors` they could in theory grow as a factor of user input/cluster state so copy constructors could in theory cause stalls or blow up memory usage if not careful. Especially because `chunked_vector` is a default container throughout the codebase I'd like to not make accidental copying possible.

I would prefer if we had a marker type `std::from_range_t` or a factory functions `chunked_vector::copy_from_range` and `chunked_vector::move_from_range`. I slightly prefer the more verbose factory function personally.

Fragmented vector/chunked vector *always* has had the copy constructor deleted: https://github.com/redpanda-data/redpanda/commit/b971a092187536494450af39a301bda8ce826abd#diff-62f13233d18f7a098dd26027d2818710895494e194b6b76622ed44ae17cd7879R48

We added the `copy` method as a compromise. I think accidental copies are enough of a footgun to catch in a review that I think it's better to force these use cases to be explicit.

bashtanov replied:
How can an _explicit_ copy constructor be accidentally called?

rockwotj replied:
You can forget to call `std::move`

bashtanov replied:
> It effectively becomes a copy constructor for chunked_vector just a lot slower as well.

It actually doesn't. The deleted constructor still wins in the resolution as it's not templated, and then fails to compile due to being deleted.

> You can forget to call std::move

You are right, it can happen in an initializer

> I would prefer if we had a marker type std::from_range_t or a factory functions chunked_vector::copy_from_range and chunked_vector::move_from_range

Given it's not going to work with the same type, would you mind if it was a tagged constructor?
The benefits of a tagged one, as opposed to a static member function are:
1) it's still a constructor so can be used in initializers, emplaces etc
2) `std::ranges::to<chunked_vector>` will call it.

rockwotj replied:
Happy with the constructor approach, but let's use different tags for copy vs move

bashtanov replied:
No tag for move, `std::from_range` for copy?

Or why would we need a tag for move at all? (Is it fine without a tag as it is in dev branch right now?)

bashtanov replied:
I've implemented the above, could you have a look @rockwotj?

rockwotj replied:
I will look tonight.

rockwotj replied:
Is this used in the PR anywhere? It looks like not. I'm OK to add it, but it seems like not adding it if it's not needed it probably the better approach for now?

bashtanov replied:
It is used in https://github.com/redpanda-data/redpanda/pull/26489/files#diff-7b346dd2edbaa2d74ef3b0cba2e2273c0a6cc3b55ad09c230b9135a378eb66a9R152

rockwotj replied:
Ah I see, LGTM
